synapse/http/server.py

```diff
@@ -532,15 +608,23 @@ def respond_with_json(
         return None
 
     if pretty_print:
-        json_bytes = encode_pretty_printed_json(json_object) + b"\n"
+        encoder = iterencode_pretty_printed_json
     else:
         if canonical_json or synapse.events.USE_FROZEN_DICTS:
-            # canonicaljson already encodes to bytes
-            json_bytes = encode_canonical_json(json_object)
+            encoder = iterencode_canonical_json
         else:
-            json_bytes = json.dumps(json_object).encode("utf-8")
+            encoder = _encode_json_bytes
 
-    return respond_with_json_bytes(request, code, json_bytes, send_cors=send_cors)
+    request.setResponseCode(code)
+    request.setHeader(b"Content-Type", b"application/json")
+    request.setHeader(b"Cache-Control", b"no-cache, no-store, must-revalidate")
+
+    if send_cors:
+        set_cors_headers(request)
+
+    producer = _JsonProducer(request, encoder, json_object)
```

Comment:
Hm, interesting. I would do
```python
producer = _JsonProducer(request, encoder(json_object))
```
i.e. `_JsonProducer` would only take an `Iterator[bytes]`.

Reply:
I went back and forth on this a bit actually! I was mostly trying to remove the need for `_encode_json_bytes`, but it is necessary to match the signature of the canonicaljson code. 😢 

Ironically that change would make it a `_ByteProducer`, not a `_JsonProducer` 👍 -- I think you're right that it is clearer though.

Reply:
See d43d309c6ec3e117913c86dd1b48cf8a4391d34b -- I think that looks quite a bit clearer. Good idea! 👍